rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+
+There are additional features that could leverage this proposal, such as:
+
+- Playback of recorded services (for example, from a rosbag [1]_)
+- Introspection of ROS actions, which are built on services
+- Validation of a live ROS system by referencing a recording from a previous session
+
+Though this proposal focuses on the core feature of introspecting requests and responses, we want to keep the design flexible so additional features like those listed above can be implemented in the future.
+
+
+Specification
+=============
+
+Publishing Service Events
+-------------------------
+
+Whenever a request or response is sent or received, a *service event* message will be published to a topic.
+Services are responsible for publishing a message when they receive a request and when they send a response.
+Likewise, clients are responsible for publishing a message when they send a request and when they receive a response.
+Therefore, we have a total of four possible events:
+
+:Request Sent:
+  Emitted from a client after sending a request to a service.
+:Request Received:
+  Emitted from a service after receiving a request from a client.
+:Response Sent:
+  Emitted from a service after sending a response to a client.
+:Response Received:
+  Emitted from a client after receiving a response from a service.
+
+Request event messages shall be published to the hidden topic ``/SERVICE_NAME/_request_event``, where ``SERVICE_NAME`` is the fully-qualified name of the service.
+Similarly, response event messages shall be published to the hidden topic ``/SERVICE_NAME/_response_event``.
+Note that this implies that services must have unique names.
+
+By publishing service event messages to predetermined topics, tools and libraries are able to subscribe to these topics to inspect the flow of data between services.
+
+Service Event Definitions
+-------------------------
+
+For each service definition, ``my/srv/Foo``, we define two new ROS message types using the ROS IDL specification [2]_:
+
+:my/msg/Foo_RequestEvent:
+  This type communicates a request sent or request received event.
+
+  .. code-block::
+
+     # Indicates this request event was emitted from a client
+     uint8 CLIENT_SENT      = 0
+
+     # Indicates this request event was emitted from a service
+     uint8 SERVICE_RECEIVED = 1
+
+     # Whether this is a CLIENT_SENT or SERVICE_RECEIVED event.
+     uint8 request_type
+
+     # Timestamp for when the event occurred (sent or received time)
+     builtin_interfaces/msg/Time stamp
+
+     # Unique identifier for the client that sent the request
+     # Note, this is only unique for the current session
+     unique_identifier_msgs/msg/UUID client_id
+
+     # Sequence number for the request
+     # Combined with the client ID, this creates a unique ID for the request
+     int64 sequence_number
+
+     # The actual request content sent or received
+     # Setting this field is optional
+     my/srv/Foo_Request request
```

Comment:
Perhaps a good way to make this optional is to make it a bounded array,
```suggestion
     my/srv/Foo_Request[<=1] request
```

Then it's easy in code to check if the request is actually set, for example

```py
if len(event_message.request) == 1:
    # Do something with the request ...
```

Reply:
Using a bounded array will reduced the wire size when not set but it will force an allocation when it's set in most bindings (e.g. it maps to a (wrapped) std::vector in cpp).
Both seem okay IMO.
Ideally we would support `optional` types (or for the case of request/response events in the same topic, we would support `enum` types).

Reply:
The reason for making the request and response members optional is to handle scenarios where the content may be very large. For example, if the service response is a large image or point cloud then users might want to skip the content but still have a log of service events. However, probably in most cases we want the request and response content and it should be set by default.

Reply:
Initially, I had read the optional field to mean "something went wrong" and this has prodded me into thinking about whether we might want to include a timeout event to capture the error case. For example, in the [example code](https://docs.ros.org/en/rolling/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client.html): 

```c++
  auto result = client->async_send_request(request);
  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Sum: %ld", result.get()->sum);
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Failed to call service add_two_ints");
  }
```
That `spin_until_future_complete` call must have some timeout built in. If that fires for any reason are we going to see the timeout in our interaction trace?

Reply:
I don't think there's currently any mechanism for a ROS service request timeout, so I don't think it's very feasible. At the moment, it's up to the application code to implement a timeout. E.g. you could choose to wait on a future with a timeout

```cpp
rclcpp::spin_until_future_complete(future, std::chrono::seconds(5));
``` 

but this is something that is operating on the _future_ object in general, nothing specific to services.